refactor(semantic): var hoisting

Dunqing opened the pull request:
close: #4323

This PR refactors the var hoisting logic to avoid inserting the same symbol into every scope that can be hosted.

Dunqing commented:
* **#4379** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4379?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Dunqing commented on tasks/coverage/parser_typescript.snap:
```diff
@@ -4639,6 +4639,19 @@ Expect to Parse: "compiler/withStatementInternalComments.ts"
  2 │ /*1*/ with /*2*/ ( /*3*/ false /*4*/ ) /*5*/ {}
    ·       ────
    ╰────
+Expect to Parse: "conformance/async/es6/asyncWithVarShadowing_es6.ts"
+
+  × Identifier `x` has already been declared
+     ╭─[conformance/async/es6/asyncWithVarShadowing_es6.ts:130:14]
+ 129 │     }
+ 130 │     catch ({ x }) {
+     ·              ┬
+     ·              ╰── `x` has already been declared here
+ 131 │         var x;
+     ·             ┬
+     ·             ╰── It can not be redeclared here
+ 132 │     }
+     ╰────
```
It is correct

https://esbuild.github.io/try/#dAAwLjIzLjAALS1taW5pZnkAYXN5bmMgZnVuY3Rpb24gZm4yNih4KSB7CiAgICB0cnkgewoKICAgIH0KICAgIGNhdGNoICh7IHggfSkgewogICAgICAgIHZhciB4OwogICAgfQp9

Boshen replied:
I remember this one! I was so confused :-/

overlookmotel replied:
What?? In NodeJS, this runs without error:

```js
try {
    throw { x: 123 };
} catch (x) {
    var x = 456;
}
```

and this throws `SyntaxError: Identifier 'x' has already been declared`:

```js
try {
    throw { x: 123 };
} catch ({ x }) {
    var x = 456;
}
```

But look what this does:

```js
try {
    throw {x: 123};
} catch (x) {
    console.log(x); // Logs {x: 123}
    var x = 456;
    console.log(x); // Logs 456
    var y = 789;
}
console.log(x); // Logs undefined (!!!)
console.log(y); // Logs 789
```

Note that the last `console.log(x)` does NOT throw `ReferenceError: x is not defined`. So `var x` has created a hoisted binding in outer scope, but it's not initialized (unlike `var y = 789;` which creates binding *and* sets its value in outer scope). Is this correct behavior??

Dunqing replied:
I can't explain why this is the case, from the example it seems to be because `CatchClause` can't change the value of the `CatchParameter` internally

Boshen replied:
Read the spec. It tried to fix for es2015.

Boshen commented:
https://github.com/oxc-project/monitor-oxc/actions/workflows/ci.yml can check PR now.

Boshen commented on crates/oxc_semantic/src/builder.rs:
```diff
@@ -67,6 +68,7 @@ pub struct SemanticBuilder<'a> {
     // to value like
     pub namespace_stack: Vec<SymbolId>,
     current_reference_flag: ReferenceFlag,
+    pub hosting_variables: FxHashMap<ScopeId, Vec<SymbolId>>,
```
`hosting` -> `hoisting`.

Boshen replied:
This state feels so weird. Shall we sit on this PR for at least 2 days before merging too quickly?

Dunqing replied:
Ok

Let me explain why need it. This state replaces add binding to every hoisting scope. Since we're only inserting bindings into the top-level hoisting scopes now. This will result in us not being able to check if a variable declared after the promoted variable has been declared or not

For example:

```js
function func() {
  {
    {
      var v = 0;
    }
    let v = 0;
  }
}
```

After hoisted, will be like this:

```js
function func() {
  var v;
  {
    {
      v = 0;
    }
    let v = 0;
  }
}
```
Should throw an error “Identifier `v` has already been declared” for `let v = 0;`. If without `hoisting_varaibles` we can't do that

Dunqing commented:
> https://github.com/oxc-project/monitor-oxc/actions/workflows/ci.yml can check PR now.

I saw `mangler` job succeed, and the `transformer` job failed but it has nothing to do with the changes to that PR
https://github.com/oxc-project/monitor-oxc/actions/runs/10034040791/job/27728055130

Boshen commented:
I'm going to stare at this PR while I'm on the plane to Sydney tomorrow.

Boshen reviewed (approved):
Rebase and merge~

Dunqing committed:
refactor(semantic): var hoisting (#4379)

close: #4323

This PR refactors the var hoisting logic to avoid inserting the same symbol into every scope that can be hosted.
